### ISSUE TITLE
too many `ContainerStatusUnknown` pod 

### ISSUE THREAD

**Issue by yylt** [kubernetes/kubernetes#122779]
### What happened?

```
openstack            grafana-69b4cc7666-2tp5p                            0/1     Init:ContainerStatusUnknown   0          59d
openstack            grafana-69b4cc7666-2w25h                            0/1     Init:ContainerStatusUnknown   0          60d
openstack            grafana-69b4cc7666-2x7r6                            0/1     Init:ContainerStatusUnknown   0          60d
openstack            grafana-69b4cc7666-46895                            0/1     Init:ContainerStatusUnknown   0          60d
openstack            grafana-69b4cc7666-4hs54                            0/1     Init:ContainerStatusUnknown   0          60d
openstack            grafana-69b4cc7666-4vt9k                            0/1     Init:ContainerStatusUnknown   0          60d
openstack            grafana-69b4cc7666-5b9l8                            0/1     Init:ContainerStatusUnknown   0          60d
openstack            grafana-69b4cc7666-5ccd6                            0/1     Init:ContainerStatusUnknown   0          60d
openstack            grafana-69b4cc7666-5nf7v                            0/1     Init:ContainerStatusUnknown   0          60d
openstack            grafana-69b4cc7666-66hvm                            0/1     Init:ContainerStatusUnknown   0          60d
openstack            grafana-69b4cc7666-6bcsq                            0/1     Init:ContainerStatusUnknown   0          60d
openstack            grafana-69b4cc7666-799w9                            0/1     Init:ContainerStatusUnknown   0          60d
openstack            grafana-69b4cc7666-7bn64                            0/1     Init:ContainerStatusUnknown   0          60d
openstack            grafana-69b4cc7666-7sbmv                            0/1     Init:ContainerStatusUnknown   0          60d
...
```

Upon reviewing the log, it appears that this issue arises when the `node.kubernetes.io/memory-pressure` condition transitions from True to False. Even if the current pressure is significant, but it will not immediately changing back to memory-pressure=True. As the documentation https://kubernetes.io/docs/concepts/scheduling-eviction/node-pressure-eviction/#node-condition-oscillation, the default transition period is 5 minutes(eviction-pressure-transition-period default). It seems that this is leading to the accumulation of Pods in the ContainerStatusUnknown state 

In practical terms,  it appears that while `node.kubernetes.io/memory-pressure` condition transition from True to False should wait for a certain period, such as 5 minutes, but `node.kubernetes.io/memory-pressure` condition transition from False to True should not involve a waiting period and should occur immediately. this is will resolved the generation of many invalid Pods.

### What did you expect to happen?

some pod are in `ContainerStatusUnknown` state, but not a large numbers.

### How can we reproduce it (as minimally and precisely as possible)?

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: stressng-memory
spec:
  replicas: 8
  selector:
    matchLabels:
      app: stress-ng
  template:
    metadata:
      labels:
        app: stress-ng
    spec:
      nodeSelector:
        kubernetes.io/hostname: node-1
      containers:
      - name: stress
        image: alexeiled/stress-ng:0.11.19-ubuntu
        command:
          - /bin/bash
          - -c
          - |
            stress-ng --vm 1 --vm-bytes 1G --timeout 120s &
            wait
            echo "start to sleep"
            sleep 1d
        resources:
          requests:
            memory: "128Mi"
            cpu: "20m"
```

make sure the node `node-8` has memory-pressre, and than wait 10 seconds create again.

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: v1.28.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.28.2
```

</details>

### Cloud provider

<details>

none 

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by pacoxu** [comment]
> To protect against oscillation, you can use the eviction-pressure-transition-period flag, which controls how long the kubelet must wait before transitioning a node condition to a different state. The transition period has a default value of 5m.

There is already a flag in kubelet configuration for you to tuning according to the requirement. In this case, you can set a 1s or 10s to fix the problem. However, as the doc says, this may cause oscillation.

1. `kubelet may not observe memory pressure right away` See https://kubernetes.io/docs/concepts/scheduling-eviction/node-pressure-eviction/#kubelet-may-not-observe-memory-pressure-right-away. 
2. default eviction-hard thresholds is using `memory.available<100Mi` which can be configured as well. In the doc, we can see `Operator wants to reserve 10% of memory capacity for system daemons (kernel, kubelet, etc.)`. 

BTW, Changing default value of evictionPressureTransitionPeriod  in v1beta1.KubeletConfiguration is a behavior change and we should be careful.

**Comment by yylt** [comment]
Hope that there is a period of time to transition from the state True to False and that it respects the parameter eviction-pressure-transition-period. However, transitioning from False to True does not require that period of time, the process is roughly as follows

![image](https://github.com/kubernetes/kubernetes/assets/34183906/cbced3e5-f567-44d2-bd5c-9ed6faad9fe8)

When set condition to false which means that the taint on node will be removed. Sometimes, a large number of pods may created on this node, but because the resources are already at a critical point, it is very easy to create pressure again. However, because it is must to wait for a period of time to transform, this period of time will result in a large number of invalid pod.

**Comment by pacoxu** [comment]
I am not sure if this is a bug or a feature. 
- If a feature,  it may be a good use case to add a new configuration from false to true. But that may needs a new design. A KEP may be needed. 

Why not change it to 2.5m? Then `true to false 2.5` + `false to true 2.5`. 

I would like to wait for suggestions from other sig node maintainers.

**Comment by SergeyKanzhelev** [comment]
@pacoxu can you summarize the issue? Is it a problem of how eviction works or how status is being reported similar to https://github.com/kubernetes/kubernetes/issues/122932#issuecomment-1907652941?

/triage needs-information

**Comment by pacoxu** [comment]
> @pacoxu can you summarize the issue? Is it a problem of how eviction works or how status is being reported similar to [#122932 (comment)](https://github.com/kubernetes/kubernetes/issues/122932#issuecomment-1907652941)?

Not the same reason with #122932.

https://github.com/kubernetes/kubernetes/issues/122779#issuecomment-1891721229 is a good summary.

> eviction-pressure-transition-period: duration for which kubelet has to wait before transitioning out of an eviction pressure condition. The default value is 5 minutes. If the time exceeds the threshold, the node is set to memory pressure or disk pressure, and then pod eviction is started.

This is as expected. Take memory pressure as an example.

### when memory is low later, @yylt  thinks the node should be back immediately.

I explained as https://kubernetes.io/docs/concepts/scheduling-eviction/node-pressure-eviction/#node-condition-oscillation.
```
To protect against oscillation, you can use the eviction-pressure-transition-period flag, which controls
how long the kubelet must wait before transitioning a node condition to a different state. The transition period
 has a default value of 5m.
```

### @yylt 's expectation is like that we can config the period from memory pressure to no pressure (and want a 0s comback)

I think this is not correct.

My workaround is to set `eviction-pressure-transition-period` to 2.5m, then the pod is evicted after 2.5m and if you want to run the pod again on this node, it can wait another 2.5 min. So the node recovers after 5m instead of default 10m in @yylt 's scanerio.

Another proposal would be make the `come-back-from-pressure` configurable.

 pressure type | status change  | kubelet eviction behavior 
 -- | -- | --
memory pressure  | from no pressure to memory pressure | wait `eviction-pressure-transition-period`  and then pod eviction is started.
memory pressure  | from memory pressure to no pressure | wait `eviction-pressure-transition-back-period`  and then pod eviction is started.

Then @yylt  can configure the second flag as he likes. It may cause oscillation.

**Comment by AnishShah** [comment]
This will need its own KEP since the suggestion is to change node status.

/kind feature
/remove-kind bug

**Comment by echalegre** [comment]
Hi guys, at the moment my cluster has some pod with status ContainerStatusUnknown. The cluster scale the pod on another node but the pod with status ContainerStatusUnknown is not deleted. So, I have to delete myself. There is anything that I can do about this?
This alerting in my monitoring tool a lot.

![image](https://github.com/kubernetes/kubernetes/assets/159057542/8b754f7f-bf81-4455-9a5b-3dad557bf6aa)

**Comment by yylt** [comment]
@echalegre  Hi, could you print more information? the pod which is `ContainerStatusUnknown` and the node which pod is scheduled on.

like below

```
kubectl get no {node} -o yaml

kubec get po -n {namespace} {pod} -oyaml
```

**Comment by RiRa12621** [comment]
Hitting a similar issue as @echalegre.

Doesn't seem sane, that unknown pods aren't cleaned up automatically.

```
$ oc get pods -A | grep Unknown | wc -l
      51
```
 
 
 That's on a 10 node cluster. The cluster is, indeed pretty full, but that shouldn't cause rather permanent issues.

Anything specific you're looking for in the node and pod details? I'd prefer not to share the whole output for $reason, but could probably do for the relevant details.

**Comment by KaustubhDhokte** [comment]
@echalegre @RiRa12621 @yylt  I hit the same issue. Too many pods stuck into ContainerStatusUnknown and Evicted after memory pressure. I set --terminated-pod-gc-threshold=1 on kube-controller-manager as per https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/#kube-controller-manager 
All these garbage pods were then cleaned up. See if this works for you guys too.
